src/blockrelay/graphene.cpp

```diff
@@ -234,14 +234,14 @@ bool CRequestGrapheneBlockTx::HandleMessage(CDataStream &vRecv, CNode *pfrom)
     // Check for Misbehaving and DOS
     // If they make more than 20 requests in 10 minutes then disconnect them
     {
-        LOCK(cs_vNodes);
         if (pfrom->nGetGrapheneBlockTxLastTime <= 0)
             pfrom->nGetGrapheneBlockTxLastTime = GetTime();
         uint64_t nNow = GetTime();
-        pfrom->nGetGrapheneBlockTxCount *=
+        pfrom->nGetGrapheneBlockTxCount =
```

Comment:
even with atomics, this may not work because: 
In thread 1 pfrom->nGetGrapheneBlockTxCount is read and put into a temporary while the rest of the expression is calculated.  Then thread 2 enters this same block and modifies pfrom->nGetGrapheneBlockTxCount.  Then thread 1 continues and overwrites pfrom->nGetGrapheneBlockTxCount with its calculated value.

To do this atomically we need to use compare_exchange_xxx.
pfrom->nGetGrapheneBlockTxCount =
            pfrom->nGetGrapheneBlockTxCount * std::pow(1.0 - 1.0 / 600.0, (double)(nNow - pfrom->nGetGrapheneBlockTxLastTime));

becomes:

double tmp = pfrom->nGetGrapheneBlockTxCount;
while (!pfrom->nGetGrapheneBlockTxCount.compare_exchange_weak(tmp, tmp * std::pow(1.0 - 1.0 / 600.0, (double)(nNow - pfrom->nGetGrapheneBlockTxLastTime))));

basically compare_exchange_weak says compare pfrom->nGetGrapheneBlockTxCount to "tmp" and assign the 2nd argument (tmp * std::pow(1.0 - 1.0 / 600.0, (double)(nNow - pfrom->nGetGrapheneBlockTxLastTime) ONLY if its equal.  If its not equal, load "tmp" with the new value of pfrom->nGetGrapheneBlockTxCount and try again.

This needs to be done in all places where *= is changed.

Reply:
done...